File: java/base/src/main/vsdk/toolkit/common/linealAlgebra/Vector3Df.java
package vsdk.toolkit.common.linealAlgebra;

import java.io.Serial;
import vsdk.toolkit.common.FundamentalEntity;
import vsdk.toolkit.common.VSDK;

public final class Vector3Df extends FundamentalEntity {
    @Serial
    private static final long serialVersionUID = 20260514L;

    private final float x;
    private final float y;
    private final float z;

    public Vector3Df() {
        this(0f, 0f, 0f);
    }

    public Vector3Df(float x, float y, float z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    public Vector3Df(Vector3Df other) {
        this(other.x, other.y, other.z);
    }

    public Vector3Df(Vector3Dd other) {
        this((float) other.x(), (float) other.y(), (float) other.z());
    }

    public Vector3Df multiply(float a) {
        return new Vector3Df(a * x, a * y, a * z);
    }

    public Vector3Df crossProduct(Vector3Df other) {
        return new Vector3Df(
            y * other.z - z * other.y,
            z * other.x - x * other.z,
            x * other.y - y * other.x
        );
    }

    public float dotProduct(Vector3Df other) {
        return x * other.x + y * other.y + z * other.z;
    }

    public Vector3Df normalized() {
        float t = x * x + y * y + z * z;
        if ( Math.abs(t) < (float) VSDK.EPSILON ) return this;
        if ( t != 0f && t != 1f ) t = (float) (1.0 / Math.sqrt(t));
        return new Vector3Df(x * t, y * t, z * t);
    }

    public float length() {
        return (float) Math.sqrt(x * x + y * y + z * z);
    }

    public Vector3Df add(Vector3Df b) {
        return new Vector3Df(x + b.x, y + b.y, z + b.z);
    }

    public Vector3Df subtract(Vector3Df b) {
        return new Vector3Df(x - b.x, y - b.y, z - b.z);
    }

    public Vector3Df withX(float nx) { return new Vector3Df(nx, y, z); }
    public Vector3Df withY(float ny) { return new Vector3Df(x, ny, z); }
    public Vector3Df withZ(float nz) { return new Vector3Df(x, y, nz); }
    public float x() { return x; }
    public float y() { return y; }
    public float z() { return z; }

    public boolean epsilonEquals(Vector3Df other) {
        return epsilonEquals(other, (float) VSDK.EPSILON);
    }

    public boolean epsilonEquals(Vector3Df other, float epsilon) {
        if ( other == null ) return false;
        if ( epsilon < 0f ) throw new IllegalArgumentException("epsilon must be >= 0");
        return Math.abs(x - other.x) <= epsilon &&
               Math.abs(y - other.y) <= epsilon &&
               Math.abs(z - other.z) <= epsilon;
    }
}

File: typescript/base/src/main/vsdk/toolkit/common/linealAlgebra/Vector3Df.ts
import { FundamentalEntity } from "../FundamentalEntity.js";
import { VSDK } from "../VSDK.js";
import { Vector3Dd } from "./Vector3Dd.js";
const f = Math.fround;
export class Vector3Df extends FundamentalEntity {
  private readonly xv:number;private readonly yv:number;private readonly zv:number;
  public constructor();public constructor(x:number,y:number,z:number);public constructor(other:Vector3Df|Vector3Dd);
  public constructor(a:number|Vector3Df|Vector3Dd=0,b=0,c=0){super();if(typeof a==="number"){this.xv=f(a);this.yv=f(b);this.zv=f(c);}else if(a instanceof Vector3Df){this.xv=a.x();this.yv=a.y();this.zv=a.z();}else{this.xv=f(a.x());this.yv=f(a.y());this.zv=f(a.z());}}
  public multiply(a:number):Vector3Df{return new Vector3Df(f(a*this.xv),f(a*this.yv),f(a*this.zv));} public crossProduct(o:Vector3Df):Vector3Df{return new Vector3Df(f(this.yv*o.zv-this.zv*o.yv),f(this.zv*o.xv-this.xv*o.zv),f(this.xv*o.yv-this.yv*o.xv));} public dotProduct(o:Vector3Df):number{return f(f(this.xv*o.xv)+f(this.yv*o.yv)+f(this.zv*o.zv));}
  public normalized():Vector3Df{let t=this.dotProduct(this);if(Math.abs(t)<f(VSDK.EPSILON))return this;if(t!==1)t=f(1/Math.sqrt(t));return this.multiply(t);} public length():number{return f(Math.sqrt(this.dotProduct(this)));} public add(o:Vector3Df):Vector3Df{return new Vector3Df(f(this.xv+o.xv),f(this.yv+o.yv),f(this.zv+o.zv));}public subtract(o:Vector3Df):Vector3Df{return new Vector3Df(f(this.xv-o.xv),f(this.yv-o.yv),f(this.zv-o.zv));}
  public withX(x:number):Vector3Df{return new Vector3Df(x,this.yv,this.zv);}public withY(y:number):Vector3Df{return new Vector3Df(this.xv,y,this.zv);}public withZ(z:number):Vector3Df{return new Vector3Df(this.xv,this.yv,z);}public x():number{return this.xv;}public y():number{return this.yv;}public z():number{return this.zv;}public epsilonEquals(o:Vector3Df|null,e=f(VSDK.EPSILON)):boolean{if(o===null)return false;if(e<0)throw new RangeError("epsilon must be >= 0");return Math.abs(this.xv-o.xv)<=e&&Math.abs(this.yv-o.yv)<=e&&Math.abs(this.zv-o.zv)<=e;}
}
